warehouse/forklift/legacy.py

```diff
@@ -693,7 +693,7 @@ def file_upload(request):
         release = (
             request.db.query(Release)
             .filter(
-                (Release.project == project) & (Release.version == form.version.data)
+                (Release.project == project) & (Release.version == str(meta.version))
```

Comment:
Not strictly necessary but this moves us closer to using the metadata as the source of truth everywhere.

Reply:
If I'm reading this correctly, then we no longer use `form.version.data` anywhere else - we're doing some validation on submission, but effectively ignoring the results. Is that intended?

Reply:
Ah, yes, I meant to remove that from the form. All the validation it does is already handled by `packaging` as well. Done in b3b7024.

Reply:
It looks like we still use the form's `version` field, but only because we deconstructed it into `meta` earlier?
https://github.com/pypi/warehouse/blob/f4d18dba7c397a7ceadf5f78db32528596e26ec1/warehouse/forklift/legacy.py#L499-L503

We no longer validate the form's `version`, but we use the version here for lookup and in other comparisons, not from the filename.

Reply:
> It looks like we still use the form's `version` field, but only because we deconstructed it into `meta` earlier?

Correct. We're still using the version in the POST body as the source of truth, until we resolve the TODO listed here and get all metadata from within the artifact.

> We no longer validate the form's version, but we use the version here for lookup and in other comparisons, not from the filename.

The version is still being validated when we're passing it to `metadata.parse`. We never used the version from the filename for any validation, the key thing this PR adds is validating that the POST data and the filename versions match.